### ISSUE TITLE
error with com.eclipsesource.json

### ISSUE THREAD

**Issue by ssoulier** [databricks/spark-redshift#279]
I am reading a redshift table. Everything is fine until a try to `collect` or `show` the result

```python
import pyspark
sc = pyspark.SparkContext("local[*]")
from pyspark.sql import SQLContext

hadoopConf = sc._jsc.hadoopConfiguration()
hadoopConf.set("fs.s3n.impl", "org.apache.hadoop.fs.s3native.NativeS3FileSystem")
hadoopConf.set("fs.s3n.awsAccessKeyId", "myAccessKey")
hadoopConf.set("fs.s3n.awsSecretAccessKey"," mySecretKey")


sql_context = SQLContext(sc)

df = sql_context.read \
    .format("com.databricks.spark.redshift") \
    .option("url", "jdbc:redshift://redshift.redshift.amazonaws.com:5439/database?user=userId&password=pass") \
    .option("dbtable", "mytable") \
    .option("tempdir", "s3n://path/to/temp") \
    .load()
```

At this stage `print df` output `DataFrame[device_id: string, server_time: bigint, field: string, value: string]`

but when I `df.show()` I have the following error :

```
Py4JJavaError: An error occurred while calling o28.showString.
: java.lang.NoClassDefFoundError: com/eclipsesource/json/Json
	at com.databricks.spark.redshift.RedshiftRelation.buildScan(RedshiftRelation.scala:134)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$8.apply(DataSourceStrategy.scala:260)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$8.apply(DataSourceStrategy.scala:260)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:303)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:302)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProjectRaw(DataSourceStrategy.scala:376)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProject(DataSourceStrategy.scala:298)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.apply(DataSourceStrategy.scala:256)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:60)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:60)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:61)
	at org.apache.spark.sql.execution.SparkPlanner.plan(SparkPlanner.scala:47)
	at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1$$anonfun$apply$1.applyOrElse(SparkPlanner.scala:51)
	at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1$$anonfun$apply$1.applyOrElse(SparkPlanner.scala:48)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:301)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:301)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:300)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:298)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:298)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$5.apply(TreeNode.scala:321)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:179)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformChildren(TreeNode.scala:319)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:298)
	at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1.apply(SparkPlanner.scala:48)
	at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1.apply(SparkPlanner.scala:48)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:78)
	at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:76)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:83)
	at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:83)
	at org.apache.spark.sql.Dataset.withTypedCallback(Dataset.scala:2558)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:1924)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2139)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:239)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:237)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:128)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:211)
	at java.lang.Thread.run(Thread.java:745)
```

eclipsesource.mini-json seems to be responsible for this issue.

**Comment by JoshRosen** [comment]
How are you depending on `spark-redshift`? Are you using `--packages` (or something similar) or are you manually adding every dependency by hand with `--jars`?

**Comment by ssoulier** [comment]
I am totally beginner with Java and Spark. So I tested almost everything before I succeeded to communicate with redshift.
First I used `--packages` like this :

```python
os.environ['PYSPARK_SUBMIT_ARGS'] = '--packages com.databricks:spark-redshift_2.10:1.1.0,com.amazonaws:aws-java-sdk:1.11.39,org.apache.hadoop:hadoop-aws:2.7.3,com.eclipsesource.minimal-json:minimal-json:0.9.4  pyspark-shell
``` 

I have this kind of message 

```
:: evicted modules:
	org.slf4j#slf4j-api;1.7.5 by [org.slf4j#slf4j-api;1.7.10] in [default]
	org.slf4j#slf4j-api;1.6.4 by [org.slf4j#slf4j-api;1.7.5] in [default]
	commons-logging#commons-logging;1.2 by [commons-logging#commons-logging;1.1.3] in [default]
	com.amazonaws#aws-java-sdk-simpleworkflow;1.11.22 by [com.amazonaws#aws-java-sdk-simpleworkflow;1.11.39] in [default]
	commons-codec#commons-codec;1.4 by [commons-codec#commons-codec;1.9] in [default]
	org.apache.httpcomponents#httpclient;4.2.5 by [org.apache.httpcomponents#httpclient;4.5.2] in [default]
	org.apache.httpcomponents#httpcore;4.2.5 by [org.apache.httpcomponents#httpcore;4.4.4] in [default]
	org.apache.avro#avro;1.7.4 by [org.apache.avro#avro;1.7.6] in [default]
	com.fasterxml.jackson.core#jackson-databind;2.2.3 by [com.fasterxml.jackson.core#jackson-databind;2.6.6] in [default]
	com.fasterxml.jackson.core#jackson-annotations;2.2.3 by [com.fasterxml.jackson.core#jackson-annotations;2.6.0] in [default]
	com.amazonaws#aws-java-sdk;1.7.4 by [com.amazonaws#aws-java-sdk;1.11.39] in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |  158  |   16  |   16  |   11  ||  147  |   0   |
	---------------------------------------------------------------------

:: problems summary ::
:::: ERRORS
	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null

	unknown resolver null
```

So I also tested to put `spark-redshift` and `amazonaws` jars in my `$SPARK_HOME/jars` directory.

Both method give the same `eclipsesource.json` error

Regarding **spark**, I downloaded spark 2.0.0 pre-build for Hadoop1.7 an later

As I said, I am totally new to Java/scala so any advice on the way I can add dependency to spark for a python usage is more than welcome.

**Comment by JoshRosen** [comment]
I can't manage to reproduce this issue. Do you see resolution messages for the `eclipsesource` JSON library further up in the log?

**Comment by ssoulier** [comment]
The following messages in the log are

```
Caused by: java.lang.ClassNotFoundException: com.eclipsesource.json.Json
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	... 48 more
```

**Comment by JoshRosen** [comment]
I still can't manage to reproduce this, so I'm going to resolve this issue as "cannot reproduce" for now. This is most likely an issue in your environment / deployment rather than a bug specific to `spark-redshift` itself, since it seems that the classes simply aren't present, so either the minimal json dependency isn't being downloaded properly or there's some other problem. It's going to be hard to debug this without more logs, etc.

**Comment by rbhatia5** [comment]
I'm running into this issue as well. Only when trying to use a read, I'm having trouble resolving the import com.eclipsesource.json.Json dependency. I'm sitting behind a corporate firewall without internet access, so maybe that might be affecting my ability to resolve this dependency.

**Comment by jpgbus** [comment]
Though the same artifacts and config yielded a successful dataframe load with just native Spark JDBC (S3 related issue), I also reproduced with:

spark 2.0.2
aws-java-sdk-1.7.4.jar
hadoop-aws-2.7.3.jar
spark-redshift_2.11-2.0.1.jar
RedshiftJDBC42-1.2.1.1001.jar

```val jdbcUrl = s"jdbc:redshift://${host}:${port}/${db}?user=${user}&password=${password}&ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory"
val dbtable = s"${schema}.${table}"

val tempS3Dir = s"s3a://spark/redshift/${table}_${System.currentTimeMillis}/"
val df = spark
  .read
  .format("com.databricks.spark.redshift")
  .option("driver", "com.amazon.redshift.jdbc42.Driver")
  .option("url", jdbcUrl)
  .option("tempdir", tempS3Dir)
  .option("dbtable", dbtable) 
.option("forward_spark_s3_credentials", true)
  .load()
  
val tempTable = s"${table}_jdbc"
df.createOrReplaceTempView(s"${tempTable}")
spark.sql(s"select * from ${tempTable} limit 10").show(1)
```

```
java.lang.NoClassDefFoundError: com/eclipsesource/json/Json
  at com.databricks.spark.redshift.RedshiftRelation.buildScan(RedshiftRelation.scala:131)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$8.apply(DataSourceStrategy.scala:260)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$8.apply(DataSourceStrategy.scala:260)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:303)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:302)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProjectRaw(DataSourceStrategy.scala:379)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProject(DataSourceStrategy.scala:298)
  at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.apply(DataSourceStrategy.scala:256)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:60)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:60)
  at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
  at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
  at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:61)
  at org.apache.spark.sql.execution.SparkPlanner.plan(SparkPlanner.scala:47)
  at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1$$anonfun$apply$1.applyOrElse(SparkPlanner.scala:51)
  at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1$$anonfun$apply$1.applyOrElse(SparkPlanner.scala:48)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:308)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:308)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:69)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:307)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:305)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$4.apply(TreeNode.scala:305)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$5.apply(TreeNode.scala:328)
  at org.apache.spark.sql.catalyst.trees.TreeNode.mapProductIterator(TreeNode.scala:186)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformChildren(TreeNode.scala:326)
  at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:305)
  at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1.apply(SparkPlanner.scala:48)
  at org.apache.spark.sql.execution.SparkPlanner$$anonfun$plan$1.apply(SparkPlanner.scala:48)
  at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
  at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:78)
  at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:76)
  at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:83)
  at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:83)
  at org.apache.spark.sql.Dataset.withTypedCallback(Dataset.scala:2572)
  at org.apache.spark.sql.Dataset.head(Dataset.scala:1934)
  at org.apache.spark.sql.Dataset.take(Dataset.scala:2149)
  at org.apache.spark.sql.Dataset.showString(Dataset.scala:239)
  at org.apache.spark.sql.Dataset.show(Dataset.scala:526)
  at org.apache.spark.sql.Dataset.show(Dataset.scala:486)
  ... 48 elided
Caused by: java.lang.ClassNotFoundException: com.eclipsesource.json.Json
  at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  ... 87 more

```

**Comment by jpgbus** [comment]
updated to ```spark-redshift_2.11-3.0.0-preview1.jar```

...same error.

**Comment by arvindkanda** [comment]
I am having this issue also. Was anyone able to resolve this error?

**Comment by arvindkanda** [comment]
It requires minimum-json package. After installing it in the spark cluster, this issue went away. You can download it from here: https://s3.amazonaws.com/emrsupport/spark/spark-redshift-databricks/minimal-json-0.9.5-SNAPSHOT.jar

**Comment by suryasev** [comment]
Following arvindkandra's suggestion, I was also able to fix this by adding the 0.9.4 version to my build script.  In gradle it looks like this: `compile group: 'com.eclipsesource.minimal-json', name: 'minimal-json', version: '0.9.4'`

**Comment by leeadh** [comment]
Hi I ran in to the same problem. (Caused by: java.lang.ClassNotFoundException: com.eclipsesource.json.Json)

I placed the jar file suggested above ( https://s3.amazonaws.com/emrsupport/spark/spark-redshift-databricks/minimal-json-0.9.5-SNAPSHOT.jar)into my hadoop spark jar directory and proceeded to run pyspark. However, it still cant work and is now giving me a 

Py4JJavaError: An error occurred while calling o107.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 1, localhost, executor driver): java.lang.NoClassDefFoundError: scala/collection/GenTraversableOnce$class.....Caused by: java.lang.NoClassDefFoundError: scala/collection/GenTraversableOnce$class

Any help?

**Comment by mprice-asapp** [comment]
I ran into the same problem.

Based off of @suryasev's solution I added "com.eclipsesource.minimal-json:minimal-json:0.9.4" to my list of dependencies while launching spark-shell and got it to work:

`spark-shell --packages com.eclipsesource.minimal-json:minimal-json:0.9.4`

**Comment by anagaraju** [comment]
I have also facing same error. But it is resolved when upgrade to minimal-json:0.9.4
